Navigation: Next button behaviour

Itxaka opened the pull request.

Commit:
Small test for the navigation next button to link it to the proper state of the steps

santiph commented on assets/app/features/upgrade/upgrade.controller.js:
```diff
@@ -59,6 +59,10 @@
                 return;
             }
+            // Prevent clicking if steps is not finished
+            if (!vm.steps.activeStep.finished) {
```
merge with previous condition

Itxaka replied:
yup, this was separated to be more clear on this pr :+1:

Itxaka replied:
done!

santiph commented on assets/app/features/upgrade/controllers/upgrade-backup.controller.spec.js:
```diff
@@ -21,15 +21,22 @@ describe('Upgrade Flow - Backup Controller', function() {
         bard.appModule('crowbarApp');
         bard.inject('$controller', '$rootScope', '$q', '$httpBackend', '$document', 'crowbarBackupFactory');
-
+        $rootScope.upgradeVm = upgradeVm;
```
```
var scope = $rootScope.$new();
scope.upgradeVm = upgradeVm
```

Itxaka replied:
Is this as to create a new scope on each test? Im guessing than using $rootScope directly means Im reusing the same scope on each test which could lead to instability on the tests that use the scope right?

santiph replied:
Well, $rootScope is being defined as the ancestor of all scopes in an angular application, but given you're injecting it in a beforeEach section (that will be executed before each test is run), I'd expect to have a clean one each time.

I couldn't find a strong argument to use a new scope instead of the $rootScope for this case.

Itxaka replied:
Thanks, done!

santiph commented on assets/app/features/upgrade/controllers/upgrade-administration-repositories-checks.controller.spec.js:
```diff
@@ -50,9 +57,9 @@ describe('Upgrade Flow - Admin Repositories Checks Controller', function () {
             '$rootScope', 'PRODUCTS_REPO_CHECKS_MAP'
         );
-
+        $rootScope.upgradeVm = upgradeVm;
```
```
var scope = $rootScope.$new();
scope.upgradeVm = upgradeVm
```

Itxaka replied:
done!

santiph commented on assets/app/widgets/crowbar-upgrade-nav/crowbar-upgrade-nav.directive.jade:
```diff
@@ -1,7 +1,7 @@
   span(aria-hidden="true")
   | Cancel Upgrade
-button.btn.btn-success.pull-right(type='button', ng-click="steps.nextStep()", ng-if="!steps.isLastStep()")
+button.btn.btn-success.pull-right(type='button', ng-click="steps.nextStep()", ng-if="!steps.isLastStep()", ng-class="{disabled : steps.activeStep.finished != true   }")
```
Can you use `ng-disabled` instead?
Also, too many spaces before closing `ng-class` attribute

Itxaka replied:
done!

Commit:
- Extend the next checks and ActiveStep.finished to all the steps
 - Enables the next button by default on the database page as there is nothing there yet
 - Updates the test specs to run with a clean $scope

Itxaka commented:
Wont pass the tests until https://github.com/crowbar/crowbar-ui/pull/74 is merged and the coverage threshold goes up.

Itxaka commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(3) S49P20 #6 Layout - Navigation bar](https://trello.com/c/LNr710jO/104-3-s49p20-6-layout-navigation-bar)

santiph commented on assets/app/features/upgrade/controllers/upgrade-administration-repositories-checks.controller.js:
```diff
@@ -69,6 +69,8 @@
                             return vm.repoChecks.checks[k].status === true
                         });
+                        // Enable next button if all the checks pass
+                        $scope.upgradeVm.steps.activeStep.finished = vm.repoChecks.valid;
```
Child controllers shouldn't have visibility over its parents. We need to find an alternative way to expose the completion of the step.

santiph reviewed (changes requested):
Child controllers shouldn't have visibility over its parents. We need to find an alternative way to expose the completion of the step.
Also, unit testing for crowbarUpgradeNav directive are missing and should be covered as part of this card.
Same thing with the unit test related to the navigation functions on the main Upgrade Controller.

BTW, have you considered moving the navigation logic into its own provider (factory?) the same way we did encapsulate the Steps sidebar?

Itxaka commented:
Better solution provided on https://github.com/crowbar/crowbar-ui/pull/77, closing this
